### PR TITLE
feat(source<T>): a function to capture all values

### PR DESCRIPTION
**Pull request by coryan**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4324)

### CODE REVIEW

File: google/cloud/testing_util/mock_sources.h [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+namespace testing_util {
+
+/**
+ * A type meeting the requirements for source<T, E> used in testing.
+ *
+ * @tparam T the value type for the source
+ * @tparam E the error type for the source
+ */
+template <class T, typename E>
+class MockSource {
```
Comment (devjgm) [inline google/cloud/testing_util/mock_sources.h:36]: Naming nit: The word "mock" here may make readers think that this object is doing some verification (it's not, is it?). Especially with our use of Googletest, readers may expect it to respond to ON_CALL() and friends.

I think "Fake" may be a more commonly used term for this type of class. WDYT?

File: google/cloud/internal/source_accumulators.h [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+namespace google {
+namespace cloud {
+inline namespace GOOGLE_CLOUD_CPP_NS {
+namespace internal {
+
+template <typename Source>
+struct AccumulateAllEventsType {
+  using type = absl::variant<std::vector<typename Source::value_type>,
+                             typename Source::error_type>;
+};
```
Comment (devjgm) [inline google/cloud/internal/source_accumulators.h:33]: I'm wondering if you actually need both the struct and the alias. Could both of these be replaced with something like the following?

```
template <typename Source>
using AccumulateAllEventsType =
    absl::variant<std::vector<typename Source::value_type>,
                              typename Source::error_type>;
```

File: google/cloud/internal/source_accumulators.h [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+ * `E` are the value type and error type in the original source. The returned
+ * future is satisfied with an error if the original source is closed with an
+ * error.
+ *
+ * @tparam Source
+ * @param source
+ * @return
+ */
+template <typename Source>
+future<accumulate_all_events_t<absl::decay_t<Source>>> accumulate_all(
```
Comment (devjgm) [inline google/cloud/internal/source_accumulators.h:54]: `accumulate_all()` doesn't match the function naming convention of our style guide. Is it intentionally different to match the style of something else?

Related question for `start()`, `on_next()` below.

File: google/cloud/testing_util/mock_sources.h (hunk 1 above)
Comment (coryan) [reply google/cloud/testing_util/mock_sources.h]: Fixed.

File: google/cloud/internal/source_accumulators.h (hunk 2 above)
Comment (coryan) [reply google/cloud/internal/source_accumulators.h]: Fixed.

File: google/cloud/internal/source_accumulators.h (hunk 3 above)
Comment (coryan) [reply google/cloud/internal/source_accumulators.h]: Fixed

> CamelCase delenda est

File: google/cloud/testing_util/mock_sources.h [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+namespace testing_util {
+
+/**
+ * A type meeting the requirements for source<T, E> used in testing.
+ *
+ * @tparam T the value type for the source
+ * @tparam E the error type for the source
+ */
+template <class T, typename E>
+class FakeSource {
```
Comment (devjgm) [inline google/cloud/testing_util/mock_sources.h:36]: Note: this file is still named `mock_sources.h`.

Comment (coryan) [reply google/cloud/testing_util/mock_sources.h]: Renamed. I expect we will have more mocks (for real mocks) eventually.

**Pull request merged**